In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv')

In [3]:
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [4]:
df.count()

Laptop          2160
Status          2160
Brand           2160
Model           2160
CPU             2160
RAM             2160
Storage         2160
Storage type    2118
GPU              789
Screen          2156
Touch           2160
Final Price     2160
dtype: int64

How many records are in the dataset?

In [7]:
len(df)

2160

2160 records in the dataset

How many laptop brands are presented in the dataset?

In [9]:
df['Brand'].nunique()

27

There are 27 unique laptop brands

How many columns in the dataset have missing values?

In [13]:
df.isna().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

From the above table, it's clear the 3 columns 'Storage type', 'GPU' and 'Screen' have missing values.

What's the maximum final price of Dell notebooks in the dataset?

In [22]:
df[df['Brand'] == 'Dell']['Final Price'].max()

np.float64(3936.0)

The maximum final price of Dell notebooks in the dataset in 3936.

Median value of Screen

Find the median value of Screen column in the dataset.
Next, calculate the most frequent value of the same Screen column.
Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
Now, calculate the median value of Screen once again.

Find the median value of Screen column in the dataset.

In [34]:
df.Screen.median()

np.float64(15.6)

Next, calculate the most frequent value of the same Screen column.

In [35]:
df.Screen.mode()

0    15.6
Name: Screen, dtype: float64

Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.

In [47]:
df.Screen = df.Screen.fillna(df.Screen.mode())

Now, calculate the median value of Screen once again.

In [39]:
df.Screen.median()

np.float64(15.6)

The median remains the same.

Sum of weights

Select all the "Innjoo" laptops from the dataset.
Select only columns RAM, Storage, Screen.
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Compute the inverse of XTX.
Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the sum of all the elements of the result?

Select all the "Innjoo" laptops from the dataset.

In [63]:
df.loc[df.Brand == 'Innjoo']

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

Select only columns RAM, Storage, Screen.

In [7]:
df[df.Brand == 'Innjoo'][['RAM','Storage','Screen']]

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

Get the underlying NumPy array. Let's call it X.

In [5]:
X = np.array(df[df.Brand == 'Innjoo'][['RAM','Storage','Screen']])

In [6]:
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [7]:
Y = X.transpose()

In [8]:
multiplication_result = np.dot(Y, X)

In [9]:
multiplication_result

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

Compute the inverse of XTX

In [14]:
inverse_XTX = np.linalg.inv(multiplication_result)

In [15]:
inverse_XTX

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

Create an array y with values [1100, 1300, 800, 900, 1000, 1100].

In [17]:
y = np.array([1100, 1300, 800, 900, 1000, 1100])

In [18]:
y

array([1100, 1300,  800,  900, 1000, 1100])

Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

In [19]:
w = np.dot(np.dot(inverse_XTX, X.transpose()), y)

In [20]:
w

array([45.58076606,  0.42783519, 45.29127938])

What's the sum of all the elements of the result?

In [21]:
w.sum()

np.float64(91.2998806299555)